### ISSUE TITLE
invalid key:  advice.detachedHead

### ISSUE THREAD

**Issue by ran-arigur** [golangci/golangci-lint#6205]
### Welcome

- [x] Yes, I'm using a binary release within 2 latest [releases](https://github.com/golangci/golangci-lint/releases). Only such installations are supported.
- [x] Yes, I've searched similar [issues on GitHub](https://github.com/golangci/golangci-lint/issues) and didn't find any.
- [x] Yes, I've read the `typecheck` section of the [FAQ](https://golangci-lint.run/docs/welcome/faq/#why-do-you-have-typecheck-errors).
- [x] Yes, I've tried with the standalone [linter](https://golangci-lint.run/docs/linters/) if available (e.g., gocritic, go vet, etc.).
- [x] I agree to follow this project's [Code of Conduct](https://github.com/golangci/golangci-lint?tab=coc-ov-file)

### How did you install golangci-lint?

Official binary

### Description of the problem

On a system where the `git` command is Git v2.52.0 (which is the latest version as of this writing), `golangci-lint custom -v` fails with messaging like this:

```
INFO Cloning golangci-lint repository             
INFO error: invalid key:  advice.detachedHead
fatal: unable to write parameters to config file 
Error: build process: clone golangci-lint: git clone --branch v2.1.1 --single-branch --depth 1 -c advice.detachedHead=false -q https://github.com/golangci/golangci-lint.git: exit status 128
Failed executing command with error: build process: clone golangci-lint: git clone --branch v2.1.1 --single-branch --depth 1 -c advice.detachedHead=false -q https://github.com/golangci/golangci-lint.git: exit status 128
```

The issue is that golangci-lint is passing `-c advice.detachedHead=false` as a single command-line argument with a space in it [[code link](https://github.com/golangci/golangci-lint/blob/3111b206282d7b2dbfb1d7e4f35b61d7228377a9/pkg/commands/internal/builder.go#L93-L97)], which means that the config key is <tt> advice.detachedHead</tt> with a leading space. (Older versions of Git silently strip out that leading space, but apparently Git v2.52.0 is stricter? Or at least, this is the case with the Git v2.52.0 installed via Homebrew on a Mac. I haven't tested with other systems, and I don't see any documentation about this change.)

This can be fixed either by passing `-c` and `advice.detachedHead=false` as two separate arguments, or to pass `-cadvice.detachedHead=false` as a single argument with no space. (In the latter case I'd suggest writing something like `"-c"+"advice.detachedHead=false"` for searchability of `advice.detachedHead`.)

Thanks in advance!

### Version of golangci-lint

<details>

```console
$ golangci-lint --version
golangci-lint has version 2.6.2 built with go1.25.3 from dc16cf43 on 2025-11-14T13:00:52Z
```

</details>

### Configuration

<details>

```console
# paste configuration file or CLI flags here
```

</details>

### Go environment

<details>

```console
$ go version && go env
# paste output here
```

</details>

### Verbose output of running

<details>

```console
$ golangci-lint cache clean
$ golangci-lint run -v
# paste output here
```

</details>

### A minimal reproducible example or link to a public repository

<details>

```go
// add your code here
```

</details>

### Validation

- [x] Yes, I've included all information above (version, config, etc.).

### Supporter

- [ ] I am a sponsor/backer through [GitHub](https://github.com/sponsors/golangci/) or [OpenCollective](https://opencollective.com/golangci-lint)

**Comment by ldez** [comment]
I checked the [git v2.52 release note](https://raw.githubusercontent.com/git/git/9a2fb147f2c61d0cab52c883e7e26f5b7948e3ed/Documentation/RelNotes/2.52.0.adoc), but I found nothing relevant about flags/parsing/clone/etc.

In all cases, the arg is not right, I will fix that: #6206

**Comment by ran-arigur** [comment]
> In all cases, the arg is not right, I will fix that: https://github.com/golangci/golangci-lint/pull/6206

Thank you so much! ❤️

> I checked the [git v2.52 release note](https://raw.githubusercontent.com/git/git/9a2fb147f2c61d0cab52c883e7e26f5b7948e3ed/Documentation/RelNotes/2.52.0.adoc), but I found nothing relevant about flags/parsing/clone/etc.

Yeah. 😕

But it definitely worked with v2.51.0 and doesn't work with v2.52.0, so _something_ must have changed; maybe an accidental change resulting from some refactoring? I'll look into raising it with the Git project; I don't think there's any reason that it _should_ support `-c KEY=VALUE` as a single argument, but given that it _did_ support it, it's probably best to continue supporting it.

**Comment by jakebailey** [comment]
I'm not sure if you reported this upstream or not, but I bisected the break to https://github.com/git/git/commit/dcecac2580ef871186fdc4e9efc87815a4ce4c66, as a part of https://github.com/git/git/commit/54fef16542ef1d83b4b99d9b5e9ffd19e1942517. This doesn't strike me as intentional.

**Comment by ran-arigur** [comment]
Thanks! Yes, I tracked it down to the same commit, by a different approach (checking out and building v2.51.0, adding debugging statements to find where the trimming was happening, then finding the commit that removed that). Your way is probably better, and either way, it's good to have independent confirmation. 🙂

I haven't reported this upstream yet, but I'll do so shortly.

**Comment by jakebailey** [comment]
I just couldn't resist git bisecting git itself; when do you ever get to do that? 😄

**Comment by jakebailey** [comment]
Anyway, I hope this comes out in a 2.6.3 or a 2.7.0 soon; I was able to downgrade my `git` for now, but on other platforms that's more challenging (especially if GHA picks it up in the base image).

**Comment by ran-arigur** [comment]
> I haven't reported this upstream yet, but I'll do so shortly.

Now done. Archive link here, if you want to follow along: https://lore.kernel.org/git/CAN1UxBvk_GJjLWd0XexRxp8FFhYozGCNcodai0eqnjrhjKEh7Q@mail.gmail.com/T/#u . (Right now the message I sent is the only one on that page, but I believe that if anyone replies it will show up there as well, as well as in the linked Atom feed.)

**Comment by ran-arigur** [comment]
By the way, @jakebailey, until this is released . . .

In my team's case, we're running `golangci-lint` from inside a Bash function, so I was able to work around this for now by adding logic to temporarily put a wrapper script named `git` on the path that just splits `-c KEY=VALUE` into two separate arguments and then delegates to the regular Git binary:

```
  local real_git
  real_git="$(which git)"

  local tmp_dir
  tmp_dir="$(mktemp --directory)"
  local PATH="$tmp_dir:$PATH"

  echo '#!/bin/bash

    args=()
    for arg in "$@" ; do
      if [[ "$arg" == "-c "* ]] ; then
        args+=(-c "${arg#-c }")
      else
        args+=("$arg")
      fi
    done

    exec "'"$real_git"'" "${args[@]}"
  ' > "$tmp_dir/git"
  chmod +x "$tmp_dir/git"
```

Obviously I don't know the details of how you're using golangci-lint, but maybe you can do something similar?

**Comment by jakebailey** [comment]
I'm personally fine (I know how to work around it), but my team is largely on Windows where such a thing is very annoying to do.

Easier to say to not upgrade git until one of the two projects fixes/releases it 😄

**Comment by jakebailey** [comment]
GitHub Actions now has the version of git which breaks things.

**Comment by ldez** [comment]
I will create a release soon (~24h).

**Comment by ccoVeille** [comment]
> > I haven't reported this upstream yet, but I'll do so shortly.
> 
> Archive link here, if you want to follow along: https://lore.kernel.org/git/CAN1UxBvk_GJjLWd0XexRxp8FFhYozGCNcodai0eqnjrhjKEh7Q@mail.gmail.com/T/#u .

Meanwhile. The discussion in git tool is still in progress. It's interesting to follow. At least it's confirmed, the change wasn't intentional.
